[WIP/RFC] Implement server_address()

equalsraf opened the pull request:
- os_serveraddress() returns a unique address for Neovim to listen
  on when no address is supplied.
- In Unix the behaviour is unchanged, it uses vim_tempname() to
  get a path in the Neovim temp folder.
- For Windows all local pipe addresses must start with \.\pipe\
  followed by a name,  the pipe name is set as nvim-PID.
- Fixes pipe support in Windows

Following d52fb89fb85984df23b093637b9a8738fdcfe18a. I think I've addressed all the comments, I just need to go back to Windows and test if I have not broken anything. ping @aktau @justinmk

equalsraf commented:
Ups clearly I should have moved that somewhere in os/.

equalsraf commented:
Windows requires an include for inttypes.h for PRIu64.

aktau commented:
> Windows requires an include for inttypes.h for PRIu64.

So does every other system, iirc.

equalsraf commented:
> So does every other system, iirc.

For some reason it was not building in windows without it. Maybe inttypes is included in os_unix.h/os_unix_defs

> Just a question, would this interfere somehow with multiple nvim's running alongside each other?

Should not interfere. For Unix it does not change the current behaviour. For windows the pipes are named nvim-\<PID\> so that each running process gets a unique address. With two running instances I get

```
C:\Users\raf\Dev\neovim\build>pipelist.exe  | grep nvim
nvim-11924                                        4               -1
nvim-1964
```

aktau commented:
> For some reason it was not building in windows without it. Maybe inttypes is included in os_unix.h/os_unix_defs

Likely. This is an ongoing problem :).

> Should not interfere.

Yea, I forgot that the environment variable would only be set for the terminal from which nvim was run.

justinmk commented:
Seems like this has overlap with https://github.com/neovim/neovim/pull/1758. Can we add a `api/vim.c:vim_get_instance_id()` in this PR that just returns `os_serveraddress()`?

equalsraf commented:
I can, but the random properties of one and the other are not the same:

- In Windows this is solely dependent on pid
- Unix seems dependent on libuv's implementation of mkdtemp

i.e. there can be no two instances of Neovim running at the same time where os_serveraddress() is the same, but it would be possible for two instances that run sequentially to have the same serveraddress - with what probability I cannot say.

The goal of #1758 was different, it was looking for something along the lines of an UUID - completely unique over time.

Also plugins can already get the value as returned by os_serveraddress, by requesting $NVIM_LISTEN_ADDRESS (e.g. eval via the api).

justinmk commented:
Ok, good points.

phmarek commented:
One more comment: `vim_get_instance_id()` also works if there are multiple instances visible on one machine. Eg via port forwarding a local plugin might connect to two Neovim instances on different machines that have both the same pid and/or pipe and/or temporary directory etc.

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -57,18 +58,36 @@ bool server_init(void)
+///
+const char* os_serveraddress(void)
+{
+  static char servername[ADDRESS_MAX_SIZE] = "";
```
What happens if `serverstop()` is called on this address?

justinmk commented:
Need a "tickle test" for unix and windows to verify that the correct format is returned. Also need a test to assert behavior after `call serverstop("...")` is called on the address.

tarruda commented:
@equalsraf is this still an issue after @splinterofchaos implemented `serverstart()`/`serverstop()`?

equalsraf commented:
@tarruda yes it is. 

1. a default server name is still being generated when serverstart takes no arguments in Windows needs to be a valid pipe name
2. server_init() is still being used - but can probably be replaced with serverstart($NVIM_LISTEN_ADDRESS)?

tarruda commented:
@equalsraf ok, can rebase these changes on latest master?

justinmk commented:
Still have questions in my comments above.

equalsraf commented:
@justinmk not forgotten (:D at least not completely) need to fix it anyway for serverstart and it should address it too.

equalsraf committed:
Implement server_address_new()

When creating a local socket/pipe (server_start()) Neovim used vim_tempname() to
generate a unique socket path. For Windows UNIX filepaths cannot be used as
pipe names (they must start with \\.\pipe\). This commit replaces the use of
vim_tempname() for server addresses with server_address_new().

server_address_new() generates unique names for local sockets/pipes - for UNIX
it uses vim_tempname(), for Windows generates names in the form

    \\.\pipe\nvim-PID-COUNTER

where PID is the current process id, and COUNTER is a static uint32_t counter
incremented with every call. This function is now used for server_start() and
server_init() when no address is available.

equalsraf commented:
Rewrote most of it, renamed as server_name() - it no longer returns a unique address, just mimics vim_tempname().

In UNIX just calls vim_tempname(). For Windows it returns \\\\.\pipe\nvim-PID-COUNTER where COUNTER is a static counter in `server_name()`.

As before the PID ensures there are no collisions with other Neovim instaces, the counter increments with each call - I have not checked for a wrap around, but neither does vim_tempname(), in both cases it should be harmless and result in a failed call to server_start.

Changed `server_start()` and `f_serverstart ()` to use this instead of `vim_tempname()`. 

@justinmk for windows `serverlist()` gives me ['\\\\.\pipe\nvim-6780-0', '\\\\.\pipe\nvim-6780-1']. `serverstop('\\.\pipe\nvim-6780-1')` closes the server as expected. `serverstart('xxxxxxx')` fails with permission denied.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,30 @@ void server_teardown(void)
+#ifdef WIN32
+  static uint32_t count = 0;
+  char template[ADDRESS_MAX_SIZE] = "";
+  if (template[0] == '\0') {
```
Why use an `if`? The condition is always true.

equalsraf replied:
Leftover, `template` used to be a static buffer with the value being initialized only once. Removing the if and the initializer.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,30 @@ void server_teardown(void)
+{
+#ifdef WIN32
+  static uint32_t count = 0;
+  char template[ADDRESS_MAX_SIZE] = "";
```
Array does not need to be initialized, if the `if` is removed.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,30 @@ void server_teardown(void)
+    snprintf(template, ADDRESS_MAX_SIZE,
+      "\\\\.\\pipe\\nvim-%" PRIu64 "-%" PRIu32, os_get_pid(), count++);
+  }
+  return vim_strsave(template);
```
`vim_strsave()` returns a `char_u*`.

equalsraf replied:
Switching to xstrdup.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,30 @@ void server_teardown(void)
+  }
+  return vim_strsave(template);
+#else
+  return (char *)vim_tempname();
```
This could return `NULL`, but `server_start()` does not allow a `NULL` argument.

equalsraf replied:
Whats the best approach here? Have server_start accept NULL and fail, or check for NULL in both cases where server_start is called.

There isn't much we can do on error so I would prefer the have server_start take NULL

equalsraf committed:
Allow server_start to accept a NULL argument

Return 1 if the endpoint argument is NULL, server_start() can get
a NULL value when using server_address_new() or vim_tempname(). Removed
the function attribute.

tarruda commented:
Since windows will not be supported by 0.1, postponing it

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,27 @@ void server_teardown(void)
+/// For other systems it is a path returned by vim_tempname().
+///
+/// This function is NOT thread safe
+char *server_address(void)
```
`server_address_create` or `server_address_generate` would be a less ambiguous name.

equalsraf replied:
Would `server_address_new()` be better?

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,27 @@ void server_teardown(void)
   GA_DEEP_CLEAR(&watchers, SocketWatcher *, close_socket_watcher);
 }
 
+/// Return unique address for local server.
```
> Generates a local server address unique to the current instance.

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -67,6 +68,27 @@ void server_teardown(void)
 
+/// Return unique address for local server.
+///
+/// In Windows this is a local pipe address in the format
```
"named pipe", not "local pipe" ?

justinmk commented:
@equalsraf Could you address (minor) comments and rebase this? It is needed for XDG.

equalsraf commented:
I've rebased this yesterday and again today, but Travis seems to be stuck.

justinmk commented:
@equalsraf Thanks. I emailed travis today, waiting for reply.
